submariner/IPSec-transport-mode.md

```diff
@@ -0,0 +1,114 @@
+
+# Cable Driver Enhancement - IPSec Transport Mode
+
+## Summary
+
+Submariner today supports three cable drivers: VXLAN, IPSec (tunnel mode) and WireGuard. The
+goal of this proposal is to outline the enhancements needed for the Submariner
+Cable Driver to enable IPSec in transport mode.
+
+For Overlay-over-IPSec in transport mode with further configuration, the benefits fall into the
+following categories:
+
+* **Performance**: Taking advantage of an overlay protocol wrapped in offloaded IPSec in transport mode can
+yield significant performance improvements. The existing IPSEC configuration performance on a pair of modern
+servers ~ 2Gbit maximum. Switching to IP GRE Tunnel protected by an offloaded transport mode IPSEC external
+“wrap” - 25Gbit. A difference of 12.5 times. Or in the case of VXLAN offloaded with an IPSec transport wrap - 9Gbit.
+* **Scale**: While it is theoretically possible to scale a system built around tunnel mode IPSEC to hundreds of nodes,
+it will be difficult to achieve reliability and stability, especially if there is a complex topology of tunnels. In
+order to achieve such scales, using other L2/L3 VPN tech combined with an optional external IPSEC protection provides
+a better approach, because it allows the use of dynamic routing protocols instead of complex static routing. Furthermore,
+it allows the use of traffic engineering and Site to Site QoS which is impossible in the case of IPSEC Tunnel mode.
+* **Debuggability**: Tunnel mode IPSEC is a debugging nightmare (especially if it is offloaded in part or in full). There
+is no visibility of what goes into the VPN and once it has entered the tunnel it is encrypted and thus unsuitable for debugging.
+This problem goes away if the VPN is using standard L2/L3 VPN technology which provides suitable interfaces on which the system
+administrator can snoop.
+
```

Comment:
I think we should cover also the following topics in the context of this EP:

- SM deployment 
	e.g: Do we need to 'tell' K8S control plane that our GW engine should be scheduled on a node with HW offload capabilities? if yes, how we should do that?

  - GW HA
	   Does it affect GW HA ?  how we should mark node as a candidate for GW deployment?

- Datapth
  Do we need to update SM ipsec NAT traversal 
  handling [1]? 
  Do we need to update/configure something at host's network stack/kernel 
 Any effects on datapath metrics? 

[1]
https://submariner.io/operations/nat-traversal/

Reply:
I think that the orchestration/deployment of a HW offloaded IPSec Submariner instance is a separate issue to the EP - the EP simply wants to add a new cable type for Overlay over IPSec in transport mode. Extending this EP beyond this is scope creep IMHO.

More info can be found here: https://libreswan.org/wiki/Cryptographic_Acceleration 
But from a first pass it looks like Libreswan is capable of automatically picking up on the acceleration capabilities and taking advantage of them when they are available

Re the NAT traversal:
 There are no changes/updates needed for NAT traversal for the Overlay over IPSec implementation we are proposing.  From the Submariner perspective the packets will appear the exact same as they do for tunnel mode. (the difference will be that we didn't encrypt an additional 20bytes (IP HDR) per packet)
 
IPSec ESP packets in transport mode can be encapped with UDP as per [RFC 3948]( https://datatracker.ietf.org/doc/html/rfc3948#section-3.2) and using the ipsec whack `--forceencaps` param in combination with `--ikeport 4500` can be used to accomplish this, you can see the below (partial) captures from a GW I used configured with VXLAN over IPSEC in transport mode:
![Screenshot 2022-04-14 at 12 37 08](https://user-images.githubusercontent.com/5012288/163383287-ed56418c-2bb8-4264-afa5-e79ba9fed547.png)



![Screenshot 2022-04-14 at 12 36 36](https://user-images.githubusercontent.com/5012288/163383280-8295d098-f2c5-43e6-867f-4fe0913a037c.png)

Reply:
1. According to EP summary one of the main benefits of this approach is performance, if the performance improvement depends on HW offload, I think we should at least specify what are the steps to consume the offload service.

I'm not saying we should have full design of the orchestration here, just to understand the bigger picture,  things like what is needed to consume the offload service (k8s scheduling? etc) and availability of offload service in different platforms.

2. Regarding NAT just wanted to make sure that the current nat discovery approach can be used with overlay + IPSec transport.

3. Not an IPSec expert, but I think we'll encrypt the same number of bytes in the current (tunnel)  mode and in the proposed mode (IPinIP overlay + transport IPSec) . right?

Reply:
From previous Experience - consuming HW offloads that aren't detected automatically through the software/app we are using generally involves the use of device plugins (to detect and advertise the available HW devices) that are vendor specific and tying them to resource pools that get dished out to pods that request them via annotations. This adds a high level of complexity to the system and the need for Submariner to lose its HW agnosticism. 

Instead, I would suggest the following: the NICs that support HW offload capabilities for Libreswan are pretty standard and available in all cloud environments. What we could document is to encourage the 'selection' of one these NICs for a node where one intends to deploy the Submariner GW and leverage much simpler technique using labels: a node label for the node that has the right NIC and for the GW pod to use a label selector (for that node label) to ensure that the GW lands on the right platform. Then rely on Libreswan to autodetect the HW features and do its thing. This uses a very simple approach - which keeps submariner away from the vendor specific approach.
(I can easily document this).

I should've clarified that I was excluding IPinIP from the discussion and focusing as VXLAN over IPSec transport mode as this is what's available in Submariner today. But yes for IPinIP over transport mode vs pure tunnel mode you will have the same number of bytes yes. 

On the NAT - there will be no changes to the current nat discovery approach for overlay over IPSec transport.